### ISSUE TITLE
Feature request for hbbft: request new validator set

### ISSUE THREAD

**Issue by varasev** [poanetwork/parity-ethereum#33]
When staking epoch begins, the `ReportingValidatorSet` contract must build a new validator set and pass a new array of validators to Parity.

At the moment, Parity expects for `ReportingValidatorSet.InitiateChange` event to be emitted as described in https://wiki.parity.io/Validator-Set.html. This event passes a new validator set in `_new_set` parameter.

So, when new staking epoch begins, something should trigger emitting of that event by `ReportingValidatorSet` contract.

I propose to have the next function in `ReportingValidatorSet` callable only by SYSTEM_ADDRESS:

```solidity
function newValidatorSet() public onlySystem {
  // ... building a new set ...
  emit InitiateChange(blockhash(block.number - 1), newListOfValidators);
}
```

Parity will call `newValidatorSet` by SYSTEM_ADDRESS on a block `N` when a new staking epoch begins (or when the previous staking epoch finishes). The function will build an array of new validators (`newListOfValidators`) and pass it into `InitiateChange` event.

Then Parity catches the event, sets the list of new validators and calls `finalizeChange` function (all these steps are described in https://wiki.parity.io/Validator-Set.html) on a block `N+1` (i.e. right after `InitiateChange` event is emitted).

But AFAIK, events emitted by system transactions are not observable (see https://wiki.parity.io/Validator-Set.html#limitations).

Is it possible to implement the described mechanism?

**Comment by varasev** [comment]
Another way (which doesn't use `InitiateChange` event) is to have the next state changing function in `ReportingValidatorSet` contract for setting a new validator set:

```solidity
function newValidatorSet() public onlySystem returns(address[]) {
  // ...
}
```

Parity will call `newValidatorSet()` by SYSTEM_ADDRESS when a new staking epoch is beginning. That function will build and return the array of new validators which will be used starting from the next block.

**Comment by afck** [comment]
It would be nice to have a Parity option that enables `SYSTEM_ADDRESS` events.
But for now I guess the getter function is easier.

**Comment by varasev** [comment]
Ok, but the `newValidatorSet` function described above is not a getter - it's a state changing function.

**Comment by afck** [comment]
Yes, unfortunately there's no way around that. (Well, except for a very complicated one that would implement off-chain asynchronous distributed key generation, which I think would come with a host of other problems. :grimacing:)

**Comment by varasev** [comment]
Ok, thanks for the clarification.

**Comment by varasev** [comment]
And one more moment: can we expect that `finalizeChange()` function (described in https://wiki.parity.io/Validator-Set.html) will be called by Parity after `newValidatorSet()` is called and the internal list of validators is updated?

**Comment by afck** [comment]
Should Parity keep track of the time, and call this function once per week, or should it leave the duration of the staking epoch to the contract and e.g. just call that function in _every_ block?

**Comment by varasev** [comment]
Parity should keep track of the time and call `newValidatorSet()` function once per week.

**Comment by varasev** [comment]
So, as far as I understand, the changing of validator set every new staking epoch will happen in the next order:

1. Parity calls `ReportingValidatorSet.newValidatorSet()` function (once per week) and receives from it a new list of validators.

2. Parity performs running of distributed key generation and writes all messages to `KeyGenHistory` contract during that.

3. After key generation completes, the new validator set retrieved from the contract on point 1 is applied to the internal set of validators in hbbft.

4. `ReportingValidatorSet.finalizeChange()` is called.

Please, correct me if I'm wrong.

**Comment by afck** [comment]
I'm afraid we _will_ need the event. Otherwise how would anyone but the current validators themselves know who the new validators are? (Or am I missing something?)

I think we just need to fix the ~bug~ feature that system calls can't produce events.

**Comment by varasev** [comment]
> I think we just need to fix the bug feature that system calls can't produce events.

Yes, that would be great. In this case the [`newValidatorSet` function](https://github.com/poanetwork/pos-contracts/blob/69a99ff1e81865e546622bbbebe786ebf7e4f7f3/contracts/ReportingValidatorSet.sol#L142) would emit `InitiateChange` event with the new list of validators.

**Comment by varasev** [comment]
Now the `newValidatorSet` function emits `InitiateChange` event: https://github.com/poanetwork/pos-contracts/commit/9b56c5130097e53b5b26bdc80c72c25079cadbd1
